### PR TITLE
Tagging System 2

### PR DESCRIPTION
**Pull request by xiendong**
Completed:
Adding tags via the add command
Delete tags from a task
Delete tags globally
Rename a tag
Show global list of tags without GUI
Validation checks for all the above commands

Not completed:
Entire test suite for Tagging system
Updating User guide and developer guide for this

### CODE REVIEW

File: src/main/java/seedu/todo/logic/commands/TagCommand.java [hunk 1]
```diff
@@ -1,70 +1,84 @@
...
     private static final String ERROR_INPUT_ADD_TAGS_REQUIRED
             = "A list of tags \"tag1, tag2, ...\" to add is required.";
     private static final String ERROR_INPUT_DELETE_TAGS_REQUIRED
             = "A list of tags \"tag1, tag2, ...\" to delete is required.";
-    private static final String ERROR_TAGS_DUPLICATED
-            = "You might have keyed in duplicated tag names.";
-    private static final String ERROR_TAGS_ILLEGAL_CHAR
-            = "Tags may only include alphanumeric characters, including dashes and underscores.";
+    private static final String ERROR_INPUT_RENAME_TAGS_REQUIRED
+            = "An existing tag name, and a new tag name is required.";
```
Comment (li-kai) [inline src/main/java/seedu/todo/logic/commands/TagCommand.java]: Phrasing is a little weird. "Please provide an existing tag and a new name for it." or something. Same with the next one.

File: src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        if (!tagsDoNotExist.isEmpty()) {
+            errorBag.put(parameterName, ERROR_TAGS_DO_NOT_EXIST + YOU_SUPPLIED
+                    + StringUtil.convertListToString(tagsDoNotExist.toArray(new String[0])));
+        }
+    }
+
+    /**
+     * Checks to ensure that tag names do not exist in the {@code tagPool}.
+     */
+    private void validateTagNamesDoNotExist(Collection<Tag> tagPool, String... tagNames) {
```
Comment (li-kai) [inline src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java:201]:
```
Set<String> tagNamesSet = Sets.newHashSet(tagNames);

boolean isContainingExistingTags = tagNamesSet.retainAll(tagPool);

if (isContainingExistingTags) {
    errorBag.put(parameterName, ERROR_TAGS_EXIST + YOU_SUPPLIED
            + StringUtil.convertIterableToString(tagNamesSet));
}
```

is sufficient

File: src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    private void validateDuplicatedNameTag(String... tagNames) {
+        if (!CollectionUtil.elementsAreUnique(Arrays.asList(tagNames))) {
+            errorBag.put(parameterName, ERROR_TAGS_DUPLICATED);
+        }
+    }
+
+    /**
+     * Checks to ensure that tag names exist in the {@code tagPool}.
+     */
+    private void validateTagNamesExist(Collection<Tag> tagPool, String... tagNames) {
```
Comment (li-kai) [inline src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java:178]:
```
Set<String> tagNamesSet = Sets.newHashSet(tagNames);

tagNamesSet.removeAll(tagPool);
if (!tagNamesSet.isEmpty()) {
    errorBag.put(parameterName, ERROR_TAGS_DO_NOT_EXIST + YOU_SUPPLIED
            + StringUtil.convertIterableToString(tagNamesSet));
}
```

then in `StringUtil`, use

```
public static String convertIterableToString( Iterable<?> iterable) {
    if (iterable == null) {
        return "";
    }
    return StringUtils.join(iterable, ", ");
}
```

**Review by li-kai — CHANGES_REQUESTED** [review]
Just some changes, but mostly looks acceptable to me.

File: src/main/java/seedu/todo/logic/commands/TagCommand.java (hunk 1 above)
Comment (xiendong) [reply src/main/java/seedu/todo/logic/commands/TagCommand.java]: Done

File: src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java (hunk 3 above)
Comment (xiendong) [reply src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java:178]: Done

File: src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java (hunk 2 above)
Comment (xiendong) [reply src/main/java/seedu/todo/model/tag/UniqueTagCollectionValidator.java:201]: Done

**Comment by xiendong** [comment]
Don't merge yet!

**Review by li-kai — APPROVED** [review]

**Pull request merged**